File: lib/domain/shared/policies/interaction_rule_policy.rb
# frozen_string_literal: true

module Domain
  module Shared
    module Policies
      class InteractionRulePolicy
        def self.view_allowed?(user, is_reference:, user_id:)
          user.admin? || user_id == user.id
        end

        def self.edit_allowed?(user, is_reference:, user_id:)
          user.admin? || (!is_reference && user_id == user.id)
        end

        # @return [Domain::Shared::ValueObjects::ReferenceIndexListFilter]
        def self.index_list_filter(user)
          mode = user.admin? ? :reference_or_owned : :owned_non_reference
          Domain::Shared::ValueObjects::ReferenceIndexListFilter.new(mode: mode, user_id: user.id)
        end

        def self.record_access_filter(user)
          Domain::Shared::ReferenceRecordAccessFilter.new(user: user, policy_module: self)
        end

        def self.normalize_attrs_for_create(user, params)
          attributes = params.to_h.symbolize_keys
          attributes.delete(:region) unless user.admin?

          is_reference = Domain::Shared::TypeConverters::BooleanConverter.cast(attributes[:is_reference]) || false

          if is_reference
            attributes[:user_id] = nil
            attributes[:is_reference] = true
          else
            attributes[:user_id] ||= user.id
            attributes[:is_reference] = false if attributes[:is_reference].nil?
          end

          attributes
        end

        def self.normalize_attrs_for_update(user, current_attrs, params)
          rule = current_attrs.to_h.symbolize_keys
          update_params = params.to_h.symbolize_keys

          update_params.delete(:region) unless user.admin?

          if update_params.key?(:is_reference)
            requested_reference = Domain::Shared::TypeConverters::BooleanConverter.cast(update_params[:is_reference]) || false
            reference_changed = requested_reference != rule[:is_reference]

            if reference_changed
              if requested_reference
                update_params[:user_id] = nil
              else
                update_params[:user_id] = user.id
              end
            end
          end

          update_params
        end
      end
    end
  end
end

File: crates/agrr-domain/src/shared/policies/interaction_rule_policy.rs
use crate::shared::attr::{AttrMap, AttrValue};
use crate::shared::reference_record_access_filter::{
    RecordAccessPolicy, ReferenceRecordAccessFilter,
};
use crate::shared::type_converters::cast_boolean_attr;
use crate::shared::user::User;
use crate::shared::value_objects::reference_index_list_filter::{
    ReferenceIndexListFilter, ReferenceIndexListMode,
};

/// Bridges [`ReferenceRecordAccessFilter`] to interaction rule policy functions.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct InteractionRuleRecordAccessPolicy;

impl RecordAccessPolicy for InteractionRuleRecordAccessPolicy {
    fn view_allowed(user: &User, is_reference: bool, record_user_id: Option<i64>) -> bool {
        view_allowed(user, is_reference, record_user_id)
    }

    fn edit_allowed(user: &User, is_reference: bool, record_user_id: Option<i64>) -> bool {
        edit_allowed(user, is_reference, record_user_id)
    }
}

pub fn record_access_filter(
    user: User,
) -> ReferenceRecordAccessFilter<InteractionRuleRecordAccessPolicy> {
    ReferenceRecordAccessFilter::new(user)
}

/// Ruby: `Domain::Shared::Policies::InteractionRulePolicy`
pub fn view_allowed(user: &User, is_reference: bool, user_id: Option<i64>) -> bool {
    let _ = is_reference;
    user.admin || user_id == Some(user.id)
}

pub fn edit_allowed(user: &User, is_reference: bool, user_id: Option<i64>) -> bool {
    user.admin || (!is_reference && user_id == Some(user.id))
}

pub fn index_list_filter(user: &User) -> ReferenceIndexListFilter {
    let mode = if user.admin {
        ReferenceIndexListMode::ReferenceOrOwned
    } else {
        ReferenceIndexListMode::OwnedNonReference
    };
    ReferenceIndexListFilter::new(mode, user.id)
}

pub fn normalize_attrs_for_create(user: &User, params: AttrMap) -> AttrMap {
    let mut attributes = params;

    if !user.admin {
        attributes.remove("region");
    }

    let is_reference = attributes
        .get("is_reference")
        .map(cast_boolean_attr)
        .unwrap_or(false);

    if is_reference {
        attributes.insert("user_id".into(), AttrValue::Null);
        attributes.insert("is_reference".into(), AttrValue::Bool(true));
    } else {
        if !attributes.contains_key("user_id") {
            attributes.insert("user_id".into(), AttrValue::Int(user.id));
        }
        if attributes.get("is_reference").is_none() {
            attributes.insert("is_reference".into(), AttrValue::Bool(false));
        }
    }

    attributes
}

pub fn normalize_attrs_for_update(
    user: &User,
    current_attrs: AttrMap,
    params: AttrMap,
) -> AttrMap {
    let rule = current_attrs;
    let mut update_params = params;

    if !user.admin {
        update_params.remove("region");
    }

    if let Some(is_ref_val) = update_params.get("is_reference") {
        let requested_reference = cast_boolean_attr(is_ref_val);
        let current_reference = rule
            .get("is_reference")
            .map(cast_boolean_attr)
            .unwrap_or(false);

        if requested_reference != current_reference {
            if requested_reference {
                update_params.insert("user_id".into(), AttrValue::Null);
            } else {
                update_params.insert("user_id".into(), AttrValue::Int(user.id));
            }
        }
    }

    update_params
}

#[cfg(test)]
mod policies_interaction_rule_policy_test_inline {
    use super::*;
    include!(concat!(env!("CARGO_MANIFEST_DIR"), "/test/shared/policies_interaction_rule_policy_test.rs"));
}
